copy/single.go

```diff
@@ -29,26 +29,41 @@ import (
 
 // imageCopier tracks state specific to a single image (possibly an item of a manifest list)
 type imageCopier struct {
-	c                          *copier
-	manifestUpdates            *types.ManifestUpdateOptions
-	src                        *image.SourcedImage
-	diffIDsAreNeeded           bool
-	cannotModifyManifestReason string // The reason the manifest cannot be modified, or an empty string if it can
-	canSubstituteBlobs         bool
-	compressionFormat          *compressiontypes.Algorithm // Compression algorithm to use, if the user explicitly requested one, or nil.
-	compressionLevel           *int
+	c                             *copier
+	manifestUpdates               *types.ManifestUpdateOptions
+	src                           *image.SourcedImage
+	diffIDsAreNeeded              bool
+	cannotModifyManifestReason    string // The reason the manifest cannot be modified, or an empty string if it can
+	canSubstituteBlobs            bool
+	compressionFormat             *compressiontypes.Algorithm // Compression algorithm to use, if the user explicitly requested one, or nil.
+	compressionLevel              *int
+	requireCompressionFormatMatch bool
 }
 
 // copySingleImageResult carries data produced by copySingleImage
 type copySingleImageResult struct {
 	manifest         []byte
 	manifestMIMEType string
 	manifestDigest   digest.Digest
+	compressionAlgorithms  []compressiontypes.Algorithm
 }
 
+func removeDuplicateCompression(algos []compressiontypes.Algorithm) []compressiontypes.Algorithm {
+    allKeys := make(map[string]bool)
+    list := []compressiontypes.Algorithm{}
+    for _, algo := range algos {
+        if _, value := allKeys[algo.Name()]; !value {
+            allKeys[algo.Name()] = true
+            list = append(list, algo)
+        }
+    }
+    return list
+}
+
+
 // copySingleImage copies a single (non-manifest-list) image unparsedImage, using c.policyContext to validate
 // source image admissibility.
-func (c *copier) copySingleImage(ctx context.Context, unparsedImage *image.UnparsedImage, targetInstance *digest.Digest) (copySingleImageResult, error) {
+func (c *copier) copySingleImage(ctx context.Context, unparsedImage *image.UnparsedImage, targetInstance *digest.Digest, requireCompressionFormatMatch bool) (copySingleImageResult, error) {
```

Comment:
(In the PR that adds the new parameter) it might be worthwhile to add a `copySingleImageOptions struct`
, primarily so that the boolean is named. OTOH it’s just one boolean right now, and `copyBlob…` is much worse, so blocking on that wouldn’t be fair.

Reply:
> OTOH it’s just one boolean right now, and copyBlob… is much worse, so blocking on that wouldn’t be fair.

Cool I'll implement this later in a different PR.